### ISSUE TITLE
[Feature Request]: OpenID Connect (oidc) relying party support

### ISSUE THREAD

**Issue by Firstyear** [grokability/snipe-it#12695]
### Is your feature request related to a problem? Please describe.

OpenID Connect is a popular alternative to SAML that has been growing in adoption an used. It is an extension of Oauth2 allowing structured user information to be provided to the relying party. 

Extending Snipe-IT to support OIDC would allow integration with a broader range of authentication providers. 

OIDC is defined https://openid.net/specs/openid-connect-core-1_0.html 

### Describe the solution you'd like

Snipe IT to allow OIDC as a parallel alternative to SAML.

### Describe alternatives you've considered

n/a

**Comment by phil-flip** [comment]
Technically, Google OAuth is already OIDC, is it not?

**Comment by Firstyear** [comment]
> Technically, Google OAuth is already OIDC, is it not?

Google Oauth is an OIDC authentication provider. This is a request that Snipe supports being a relying party who consumes OIDC from an authentication provider such as google.

**Comment by olivierlambert** [comment]
+1 on this, that would be wonderful to get SnipeIT login via an SSO portal in OIDC (like Keycloak). OIDC is pretty simple to configure as an admin :) (unlike SAML and such)

**Comment by Blackclaws** [comment]
I have to add that its not just simpler to setup its also much less prone to misconfiguration as the relying party can pull the configuration from the attesting party. Definitely +1 on this request.

**Comment by mustard123** [comment]
> +1 on this, that would be wonderful to get SnipeIT login via an SSO portal in OIDC (like Keycloak). OIDC is pretty simple to configure as an admin :) (unlike SAML and such)

Keycloak would be amazing and is absolutely needed

**Comment by lucboj** [comment]
+1. Would be great to have an integration with Keycloak (OIDC ) 👍

**Comment by yavorivanov-cw** [comment]
+1, Currently snipe-it is our only SAML client in keycloak, Would be nice to have it with OIDC ❤️

**Comment by soymgomez** [comment]
+1! Snipe-IT need this!

**Comment by allebone** [comment]
Much more secure too! Great add if it can be done! 

Keycloak also, here.

**Comment by therealladiesman217** [comment]
Would be down to help implementing. Im looking for a FOSS project to contribute to, to get a track record in opensource development.
We use snipeit in the company and would love to see oidc support.

As you already consider Integration, may i ask if there is slready something in work?

**Comment by snipe** [comment]
We don't have anything for OpenID support in the works presently, no. Have at ye!

**Comment by scarpentier** [comment]
Willing to put a bounty on this

**Comment by aozima** [comment]
>Google Oauth is an OIDC authentication provider. This is a request that Snipe supports being a relying party who consumes OIDC from an authentication provider such as google.

💡So, wen can edit `vendor/laravel/socialite/src/Two/GoogleProvider.php` for custom Oauth2 Provider✅。

---

`app/Models/Setting.php`

```diff
           'thumbnail_max_h'                     => 'numeric|max:500|min:25',
-          'google_client_id'                    => 'nullable|ends_with:apps.googleusercontent.com'
```

---

`app/Http/Controllers/GoogleAuthController.php`

```diff
         $user = User::where('username', $socialUser->getEmail())->first();
+        if(!$user)
+        {
+            $user = User::where('email', $socialUser->getEmail())->first();
+        }
 
         if ($user) {
             Log::debug('Google user '.$socialUser->getEmail().' found in Snipe-IT');
```

---

`vendor/laravel/socialite/src/Two/GoogleProvider.php`

```diff
     protected function getAuthUrl($state)
     {
-        return $this->buildAuthUrlFromBase('https://accounts.google.com/o/oauth2/auth', $state);
+        return $this->buildAuthUrlFromBase('https://xxx.com/login/oauth/authorize', $state);
     }

     protected function getTokenUrl()
     {
-        return 'https://www.googleapis.com/oauth2/v4/token';
+        return 'https://xxx.com/api/login/oauth/access_token';
     }

     protected function getUserByToken($token)
     {
-        if ($this->isJwtToken($token)) {
-            return $this->getUserFromJwtToken($token);
-        }
-
-        $response = $this->getHttpClient()->get('https://www.googleapis.com/oauth2/v3/userinfo', [
-            RequestOptions::QUERY => [
-                'prettyPrint' => 'false',
-            ],
-            RequestOptions::HEADERS => [
-                'Accept' => 'application/json',
-                'Authorization' => 'Bearer '.$token,
-            ],
+        $response = $this->getHttpClient()->get('https://xxx.com/api/userinfo', [
+            RequestOptions::QUERY => ['access_token' => $token],
         ]);
 
         return json_decode((string) $response->getBody(), true);
     }
```

**Comment by tilllt** [comment]
I hope this feature request will get some traction. I just tried to follow the Setup Guide for Authentik SSO in conjunction with Snipe-It and i just had to give up. SAML is way more complicated than OIDC, thats for sure, but we have other Apps authenticating via SAML. What makes it insanely complex is that Snipe-IT cant create users on Login via SAML, so on top of SAML we are supposed to configure LDAP. Thats just too much, its 2026 guys... OIDC is the way, alternatively enable SAML user provision.

**Comment by olivierlambert** [comment]
@aozima with your workaround/change you can add any OIDC? Does it provision the user too on first login?

**Comment by snipe** [comment]
That is not a viable workaround as you should never be editing vendor files. All of your changes will be blown out the next time the library dependency gets updated.

**Comment by olivierlambert** [comment]
Good advice 🙂  What would you prefer:
1. An external contribution via a PR
2. Doing it directly (and some bounty on it)

**Comment by snipe** [comment]
I'd be willing to consider a PR, but part of the issue here is that different scenarios will have different expectations. Some folks would expect a user who exists in the directory to be created on login if they don't already exist in Snipe-IT and some folks likely want the exact opposite of that. 

There are some existing OIDC libraries out there, but we haven't had time to vet them properly, and a lot of them require env vars to work, which would make it harder to use for our customers (since they don't have access to the env) which means we'd need to add them as fields in the settings table and override the configs in those libraries. :-/

**Comment by aozima** [comment]
> Some folks would expect a user who exists in the directory to be created on login if they don't already exist in Snipe-IT and some folks likely want the exact opposite of that.

Add an auto-register config option。 （default disable）

**Comment by snipe** [comment]
Yep, and that then creates yet another path that we have to secure and test in perpetuity.

**Comment by aozima** [comment]
>that then creates yet another path that we have to secure and test in perpetuity.

I think: only provide custom configuration, and test it with 1 or 2 well-known providers, only sync base & required fields.

**Comment by olivierlambert** [comment]
I agree it's not trivial. I remember we had the same questions in Xen Orchestra 2 years ago. But we never looked back since we implemented it: this unlocked many customers. OIDC is even more popular now, so I think (just my 2 cents) it's entirely worth the work. Obviously, your business is probably very different (more SaaS?).

And also I understand that even if someone does it (via a PR), the burden of maintenance is also on you: writing a feature is a thing, maintaining it is another.

So I don't have a simple solution to give 😆  But I'm happy to try to contribute if that makes sense and helps others in here to use it.

**Comment by snipe** [comment]
@aozima 

> It's not for business.

I don't know what you mean.

@olivierlambert 

> Obviously, your business is probably very different (more SaaS?).

We do hosting and also paid support, and while we do allow customers to override *some* env var settings, it's a manual process we have to perform, so it's just not super user-friendly. We allow users to override a lot of things via the Settings section, but each time we add a new setting column, it hurts a little bit - and as we expand out the company-specific settings, this will only get exponentially worse. 

We'll keep playing around with it though. Most of our users have had no problems setting up SAML/LDAP/SCIM, or if they have any issues, it's usually just a field mapping problem.

**Comment by olivierlambert** [comment]
Okay let us know if you want guinea pigs, happy to test something on my self hosted instance when/if you want to move forward 👍

**Comment by tilllt** [comment]
> We'll keep playing around with it though. Most of our users have had no problems setting up SAML/LDAP/SCIM, or if they have any issues, it's usually just a field mapping problem.

Like i said: we have other Apps authenticating via SAML, but i just got rid of LDAP and i will not get it back from the dead for Snipe-IT only :)

**Comment by snipe** [comment]
> but i just got rid of LDAP and i will not get it back from the dead for Snipe-IT only :)

I'd never ask that - we do have to support as many things as we can though, because lots of orgs (especially schools) are still using LDAP, etc. We generally recommend SCIM+SAML for the best of both worlds, but large ships turn slowly and not everyone has SCIM available.

**Comment by tilllt** [comment]
Ok, i managed to setup Authentik with SAML and SCIM & the users are provisioned. (For anyone trying to do this in Authentik: you need to setup SCIM and SAML, use SAML as Provider and SCIM as Backchannel-Provider, do NOT map Groups or login will fail)

In terms of usability, without support for Groups, the advantage of user provisioning via SCIM is low. I dont need to create the user, which is good. But i still need to enter the Snipe-IT Admin UI every time, to assign permissions / groups to the provisioned user...

**Comment by royvandongen** [comment]
Is there already any form of rc available for us to test OIDC (against Keycloak?) Also when adding OIDC it would be really nice if we can have the internal user automaticly created on succesfull login. This is currently not the case with SAML for reasons i guess

**Comment by allebone** [comment]
> succesfull

In other products using OIDC this is called "JIT - Just In Time Provisioning" just case someone is looking for a comparison element. 

To add: config using well-known URL is also a smart ask. You only need 3 items to configure if one of them is Well Known.

**Comment by it-management-stadtbetrieb** [comment]
+1. Would be great to have an integration with Keycloak

**Comment by schewara** [comment]
I just want to add some of my thoughts to this whole discussion, as trying to use SAML (with Keycloak) has been super painful since day one and after the last Keycloak update now again completely broken.


Monkey patching some existing upstream functionality, as suggested in https://github.com/grokability/snipe-it/issues/12695#issuecomment-3538536472 is the worst thing someone can do.

In my opinion, integrating providers from https://github.com/laravel/socialite and contributing related fixes directly upstream would be a much better option.

Other frameworks that provide SSO integration usually support generic OAuth or OIDC options, which would eliminate the need of maintaining an ever growing list of Identity providers.

Feel free to chime in and upvote the discussion at https://github.com/SocialiteProviders/Providers/discussions/1428 to get this idea going.


On the user provisioning topic, I totally agree that the JIT user provisioning is very convenient, but causes other problems down the road, especially during off-boarding, like

- users can still login with a password / personal access token / API key even when the SSO (SAML or OIDC) user is disabled, unless there is a setting preventing this from happening.
- cleanup of user can not be done centrally and one has to log in again to deactivate and remove the user properly from the instance.

I think SCIM is actually the best solution for this and Group support would be exceptionally great for having proper permissions during user creation out of the box.


<blockquote><img src="https://opengraph.githubassets.com/ded29a3c90d7b700424ac8aa7cdb1e253ee9bacd45be31b4748b65ca0929b8e4/laravel/socialite" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/laravel/socialite">GitHub - laravel/socialite: Laravel wrapper around OAuth 1 & OAuth 2 libraries.</a></strong></div><div>Laravel wrapper around OAuth 1 & OAuth 2 libraries. - laravel/socialite</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/bb09cfedefc58e02afdd5acba72cbc29b9b101c4d2496908ffd07bce84b5a784/SocialiteProviders/Providers/discussions/1428" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/SocialiteProviders/Providers/discussions/1428">Why no generic protocol based provider like OIDC, OAuth, ...? · SocialiteProviders Providers · Discussion #1428</a></strong></div><div>As the list of providers will probably grow, I think it would be good to also have generic providers available to reduce the maintenance overhead and allow friction-less integration of other provid...</div></blockquote>

**Comment by restena-sw** [comment]
It is maybe worth mentioning that the standardisation body behind SAML (OASIS) has disbanded the working group, and no updates will be done to the standard. With PQC on the doorstep, the protocol will become more irrelevant as time goes by.
And OIDC is for many use cases the designated replacement.
FWIW, we have Snipe-IT deployed using its SAML integration, and it works great (once you configured it) but we have more and more software pieces moving to OIDC, and I would be happy to migrate off of SAML also for Snipe-It.
